include/pybind11/stl.h

```diff
@@ -192,6 +223,33 @@ inline std::ostream &operator<<(std::ostream &os, const handle &obj) {
 
 NAMESPACE_END(pybind11)
 
+#ifdef __has_include
+
+// std::optional
+#  if __has_include(<optional>)
+#    include <optional>
+NAMESPACE_BEGIN(pybind11)
+NAMESPACE_BEGIN(detail)
+template<typename T> struct type_caster<std::optional<T>>
+    : public optional_caster<std::optional<T>> {};
+NAMESPACE_END(detail)
+NAMESPACE_END(pybind11)
```

Comment:
Very subjective opinion: it may be nice to use the more compact `namespace pybind11 { namespace detail {` and `}}` here. Just to improve the signal-to-noise ratio.